vowpalwabbit/mwt.cc

```diff
@@ -210,7 +210,11 @@ void save_load(mwt& c, io_buf& model_file, bool read, bool text)
     policy_data& pd = c.evals[policy];
     if (read) msg << "evals: " << policy << ":" << pd.action << ":" << pd.cost << " ";
     bin_text_read_write_fixed_validated(
-        model_file, reinterpret_cast<char*>(&c.evals[policy]), sizeof(policy_data), read, msg, text);
+        model_file, reinterpret_cast<char*>(&c.evals[policy].cost), sizeof(double), read, msg, text);
```

Comment:
interesting that this got surfaced in this PR

Reply:
This issue was somewhat mysterious. I saw there were valgrind errors in the current version on master which werent showing up in the python script, not sure where that started. Splitting it up into the three members of the struct gets rid of the issue.

Reply:
does any CI pipeline catch the valgrind errors you are talking about?

Reply:
Yea I actually just tried running the python test locally on master:
`python3 run_tests.py -f -j 2 -t 123 --include_flatbuffers --valgrind --clean_dirty`
and it is throwing an error. Not sure why the CI isn't seeing this

Reply:
ok we should really really fix that then, can you open a tracking issue with some details?

Reply:
https://github.com/VowpalWabbit/vowpal_wabbit/issues/3445

Reply:
I'm pretty sure what is going on here is that the struct itself has padding in some environments and is tightly packed elsewhere. The compiler version in CI probably just doesn't do it.

The valgrind issue repros for me locally, and so it is complaining that the packing bytes are understandably uninitialized. So, by splitting it up and writing each member it will always work.

You can force tight packing with `__attribute__ ((packed))` and doing so fixes the valgrind issue. The size of the struct when tightly packed is 13 as you would expect and without it is 16, so theres 3 padding bytes in there.

For structures we serialize we need to be careful about this memory alignment because if one system uses padding and the other doesn't then the model just wont load. It's a good thing this was found! I think the right fix in this case is to serialize the members separately as was done, but I'm not sure how we can check for this in future other than through valgrind

Reply:
I updated the issue above noting that its an issue with struct padding. I'll make a separate PR trying to find other places where this issue's occurring

Reply:
@jackgerrits should we be determining the packing with the attribute you suggested? If we just initialize what we need to won't that be enough?